scanpy/tools/_marker_gene_overlap.py

```diff
@@ -0,0 +1,251 @@
+
+        if adj_pval_threshold < 0:
+            logg.warn('`adj_pval_threshold` was set below 0. '
+                      'Threshold will be set to 0.')
+            adj_pval_threshold = 0
+
+        if adj_pval_threshold > 1:
+            logg.warn('`adj_pval_threshold` was set above 1. '
+                      'Threshold will be set to 1.')
+            adj_pval_threshold = 1
+
+        if top_n_markers is not None:
+            logg.warn('Both `adj_pval_threshold` and `top_n_markers` is set. '
+                      '`adj_pval_threshold` will be ignored.')
+            
+    if top_n_markers is not None:
+        if top_n_markers < 1:
+            logg.warn('`top_n_markers` was set below 1. '
+                      '`top_n_markers` will be set to 1.')
+            top_n_markers = 1
+            
+    # Get data-derived marker genes in a dictionary of sets
+    data_markers = dict()
+    cluster_ids = adata.uns[key]['names'].dtype.names
+
+    for group in cluster_ids:
+
+        if top_n_markers is not None:
+            n_genes = min(top_n_markers, adata.uns[key]['names'].shape[0])
+            data_markers[group] = set(adata.uns[key]['names'][group][:n_genes])
+
+        elif adj_pval_threshold is not None:
+            n_genes = (adata.uns[key]['pvals_adj'][group] < adj_pval_threshold).sum()
+            data_markers[group] = set(adata.uns[key]['names'][group][:n_genes])
+
+            if n_genes == 0:
+                logg.warn('No marker genes passed the significance threshold of {} '
+                          'for cluster {!r}.'.format(adj_pval_threshold, group))
+
+        else:
+            data_markers[group] = set(adata.uns[key]['names'][group])
+
+    # Find overlaps
+    if method == 'overlap_count':
+        marker_match = _calc_overlap_count(reference_markers, data_markers)
+
+        if normalize == 'reference':
+            # Ensure rows sum to 1
+            marker_match = marker_match/marker_match.sum(1)[:,np.newaxis]
+            marker_match = np.nan_to_num(marker_match)
+
+        elif normalize == 'data':
+            #Ensure columns sum to 1
+            marker_match = marker_match/marker_match.sum(0)
+            marker_match = np.nan_to_num(marker_match)
+            
+    elif method == 'overlap_coef':
+        marker_match = _calc_overlap_coef(reference_markers, data_markers)
+
+    elif method == 'jaccard':
+        marker_match = _calc_jaccard(reference_markers, data_markers)
+        
+    #Note:
+    # Could add an 'enrich' option here (fisher's exact test or hypergeometric test),
+    # but that would require knowledge of the size of the space from which the reference
+    # marker gene set was taken. This is at best approximately known.
+        
+    # Create a pandas dataframe with the results
+    marker_groups = list(reference_markers.keys())
+    clusters = list(cluster_ids)
+    marker_matching_df = pd.DataFrame(marker_match, index=marker_groups, columns=clusters)
```

Comment:
Currently, we cannot write DataFrames using `adata.write`... :(

What I would do, is add an `inplace` parameter that defaults to `False`. If users set it to `True`, they get a `NotImplementedError`.  Then, instead of adding the `marker_matching_df` `inplace` to `adata`, you simply return it.

At some point, we will be able to store dfs within `.h5ad`. Then we can also implement the `inplace` version of this function.

@flying-sheep Any chance that you would have fun implementing this. It's just some reorganization fo the code as I implemented the functionality already for `.obs` and `.var` via https://github.com/theislab/anndata/blob/473413a76ebe741754b63558c4c5a34861451c82/anndata/base.py#L193-L222

It could be a 30 min job. Upon write, each dataframe is transformed to a record array, upon read, it's transformed back to a dataframe.

Having this would also allow a much nicer return value of `rank_genes_groups`...